CMakeLists.txt

```diff
@@ -567,23 +579,38 @@ else()
 endif()
 
 target_compile_features(torrent-rasterbar PUBLIC cxx_std_11)
-
-target_compile_definitions(torrent-rasterbar PUBLIC $<$<CONFIG:Debug>:TORRENT_DEBUG>)
-
-target_compile_definitions(torrent-rasterbar PRIVATE TORRENT_BUILDING_LIBRARY)
+target_compile_definitions(torrent-rasterbar
+	PUBLIC
+		$<$<CONFIG:Debug>:TORRENT_DEBUG>
+		$<$<CONFIG:Debug>:TORRENT_USE_ASSERTS>
+	PRIVATE
+		TORRENT_BUILDING_LIBRARY
+)
+target_link_libraries(torrent-rasterbar
+	PUBLIC
+		Threads::Threads
+)
 
 if (libiconv)
 	find_package(Iconv REQUIRED)
+	list(APPEND _package_dependencies "Iconv")
 	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_ICONV)
-	target_include_directories(torrent-rasterbar SYSTEM PUBLIC ${ICONV_INCLUDE_DIR})
+	target_include_directories(torrent-rasterbar PUBLIC ${ICONV_INCLUDE_DIR})
+	target_link_libraries(torrent-rasterbar PRIVATE ${ICONV_LIBRARIES})
 endif (libiconv)
 
 if (encryption)
 	if(NOT DEFINED OPENSSL_INCLUDE_DIR OR NOT DEFINED OPENSSL_LIBRARIES)
 		FIND_PACKAGE(OpenSSL REQUIRED)
 	endif()
-	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_OPENSSL TORRENT_USE_LIBCRYPTO)
+	list(APPEND _package_dependencies "OpenSSL")
 	target_include_directories(torrent-rasterbar PUBLIC ${OPENSSL_INCLUDE_DIR})
+	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_OPENSSL TORRENT_USE_LIBCRYPTO)
+	if (_libtorrent_additional_exports)
+		target_link_libraries(torrent-rasterbar PUBLIC ${OPENSSL_LIBRARIES})
```

Comment:
shouldn't this be specifying the dependency on proper targets instead? like "OpenSSL::SSL OpenSSL::Crypto", and remove the call to `target_include_directory()` above, where the include directory is added manually.

Reply:
also, it doesn't seem like `find_package()` works to find openssl from homebrew on mac. But I suppose that's a bug/missing feature in cmake itself.

Reply:
something like this
```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -600,16 +600,13 @@ if (libiconv)
 endif (libiconv)

 if (encryption)
-       if(NOT DEFINED OPENSSL_INCLUDE_DIR OR NOT DEFINED OPENSSL_LIBRARIES)
-               FIND_PACKAGE(OpenSSL REQUIRED)
-       endif()
+       find_package(OpenSSL REQUIRED)
        list(APPEND _package_dependencies "OpenSSL")
-       target_include_directories(torrent-rasterbar PUBLIC ${OPENSSL_INCLUDE_DIR})
        target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_OPENSSL TORRENT_USE_LIBCRYPTO)
        if (_libtorrent_additional_exports)
-               target_link_libraries(torrent-rasterbar PUBLIC ${OPENSSL_LIBRARIES})
+               target_link_libraries(torrent-rasterbar PUBLIC OpenSSL::SSL OpenSSL::Crypto)
        else()
-               target_link_libraries(torrent-rasterbar PRIVATE ${OPENSSL_LIBRARIES})
+               target_link_libraries(torrent-rasterbar PRIVATE OpenSSL::SSL OpenSSL::Crypto)
        endif()
 else()
        target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_DISABLE_ENCRYPTION)
```

except, I think the public headers of libtorrent may include openssl headers, so the link probably has to be unconditionally `PUBLIC`.

Reply:
Yeah, imported targets (OpenSSL::SSL OpenSSL::Crypto here) would make a cleaner code, of course. I was puzzled by the fact, already mentioned by you just above, that headers are always public dependency while the library is either private or public.  When you say let's make them all unconditionally public, I will make the change.

CMake needs a [hint](https://github.com/qbittorrent/qBittorrent/blob/master/.travis.yml#L154) to find the openssl brew package. I guess the brew package installs it in an uncommon location.

The canonical way for CMake (and as far as I understand this is what package maintainers got used to) is to make the `find_package()` call  first (without `REQUIRED` argument) and enable or disable option basing on the outcome. CMake provides a standard way to make `find_package()` fail for a given package name. I was looking into the sources to learn how do OpenSSL headers are used, and found that GCrypt is also an option. Thus I would replace the `option(encryption ...)` call with option `find_package()` calls for OpenSSL and then GCrypt (or in the opposite order; don't know which one is preferred).

Reply:
while looking at this earlier I noticed that enabling encryption (bittorrent protocol obfuscation) and linking against openssl are tied together here. They shouldn't be and there's no reason for them to be.

I don't think it's important to be able to disable encryption (but that's just controlled by the define `TORRENT_DISABLE_ENCRYPTION`). linking against openssl enables HTTPS support. HTTPS is not supported by gcrypt (it's up to boost.asio). The gcrypt option is just to use gcrypt instead of the built-in functions for RC4 and SHA1.

I was going to make a separate patch once this lands, but if you're up for it, I think renaming this option to just openssl or something like that would make sense. And have it not affect the encryption settings.

Reply:
So, I should then make the `encryption`option to control the `TORRENT_DISABLE_ENCRYPTION` flag only, and the `find_package(OpenSSL)` to be independent? Similarly, the `libiconv` option is also useless and can be replace by `find_package()`. Then we get optional dependencies and it would make sense to notify user at the end of configure stage about results. CMake provides a [standard way](https://cmake.org/cmake/help/latest/module/FeatureSummary.html) for that. But then it makes sense to put all the options into that, which will be quite a big change and let's leave it for another PR, please (I already have those changes adding `feature_summary()` in my local repo)?

Reply:
There should be options for all optional dependencies. Automagic dependencies are annoying for Linux package maintainers.

Reply:
Disagree. `find_package()` call is a nice way to signal optional dependence. In case of CMake, I think we can follow the style used by the KDE project, which is probably the biggest CMake consuming community, and package maintainers know what to do in that case very well. Also, time to time I observe Gentoo maintainers insist on this very model in KDE bugzilla/phabrikator.

Reply:
I see, I did not realize there is a standard way to tell `find_package` not to search for a particular package. That's fine then.